src/datachain/lib/dc/storage.py

```diff
@@ -79,40 +102,73 @@ def from_storage(
     cache = session.catalog.cache
     client_config = session.catalog.client_config
 
-    list_ds_name, list_uri, list_path, list_ds_exists = get_listing(
-        uri, session, update=update
-    )
+    uris = uri if isinstance(uri, (tuple, list)) else [uri]
+
+    if not uris:
+        raise ValueError("No URIs provided")
 
-    # ds_name is None if object is a file, we don't want to use cache
-    # or do listing in that case - just read that single object
-    if not list_ds_name:
-        dc = from_values(
+    storage_chain = None
+    listed_uris = set()
+    file_values = []
+
+    for single_uri in uris:
+        update_for_uri = update
+        if update:
+            # If update is True, we need to check if the URI has already been listed
+            # If it has, we don't need to update the listing
+            _, lst_uri, _ = parse_listing_uri(os.fsdecode(single_uri), client_config)
+            if lst_uri in listed_uris:
```

Comment:
should it be `not in` at the very least? 

do we need a test for this then?

Reply:
I think we need a test for deduplication logic ...

Reply:
Added a test